src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -622,80 +636,116 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
             _incomingBuffer.Discard(frameHeader.Length);
         }
 
-        private async ValueTask AcquireWriteLockAsync()
+        private async Task StartWriteAsync(int writeBytes, CancellationToken cancellationToken = default)
         {
-            await _writerLock.WaitAsync().ConfigureAwait(false);
+            await AcquireWriteLockAsync(cancellationToken).ConfigureAwait(false);
 
-            // If the connection has been aborted, then fail now instead of trying to send more data.
-            if (IsAborted())
+            try
             {
-                throw new IOException(SR.net_http_invalid_response);
-            }
-        }
+                // If there is a pending write that was canceled while in progress, wait for it to complete.
+                if (_inProgressWrite != null)
+                {
+                    await _inProgressWrite.ConfigureAwait(false);
+                    _inProgressWrite = null;
+                }
```

Comment:
What happens if an in-progress write fails.  Could that cause every subsequent write attempt to get here, see that _inProgressWrite is non-null, and then await it and throw the same exception, with no one ever getting past this point?  I'm wondering if instead this should be:
```C#
Task inProgressWrite = _inProgressWrite;
if (inProgressWrite != null)
{
    _inProgressWrite = null;
    await inProgressWrite.ConfigureAwait(false);
}
```

Reply:
The situation you described is possible, but I think it's actually the desired behavior. Once a write fails we're going to have to kill the whole connection, so no other requests should get through. The first exception thrown should eventually lead to the connection being torn down, so we won't be stuck in this state indefinitely.